src/PureSMT/Process.hs

```diff
@@ -43,6 +43,9 @@ initZ3instance dbg = do
 
   return s
 
+freeZ3Instance :: Solver -> IO ()
+freeZ3Instance s = void $ command s $ List [Atom "exit"]
```

Comment:
Nobody seems to be calling this function.

Also, apparently destroying the context requires a call to `Z3_del_context`. I don't know if sending the exit command is necessary.

Reply:
Yeah I figured Z3 would call `Z3_del_context` by itself when evaluating `(exit)` but it was just a guess.

Reply:
I tried using
```
let ctx = context solver
 in [CU.exp| void { Z3_del_context($(Z3_context ctx)) } |]
```
but I'm getting a segmentation fault at runtime. I tried looking into what `(exit)` does but the [SMTLib documentation](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf#subsection.4.2.1) doesn't say much. I think it's fair to stick with `(exit)` at least for now: in the worst case `freeZ3Instance` just won't do anything and the memory leak won't be a serious problem because Pirouette is not used as a library yet.

Reply:
Ok so 
```
printf("%s", Z3_eval_smtlib2_string(ctx, "(assert (< 0 -1))"));
printf("%s", Z3_eval_smtlib2_string(ctx, "(check-sat)"));
printf("%s", Z3_eval_smtlib2_string(ctx, "(exit)"));
printf("%s", Z3_eval_smtlib2_string(ctx, "(check-sat)"));
```
outputs
```
unsat
unsat
```
so I guess `(exit)` doesn't do anything here.

Reply:
> but I'm getting a segmentation fault at runtime.

This looks to me the result of [src/PureSMT.hs](https://github.com/tweag/pirouette/blob/219689f1d5ecd9b487f40ffa85fe10a1398bc195/src/PureSMT.hs#L88) reusing the context after it has been destroyed.

Also, the use of `unsafePackCString` could interfere with this too.

Reply:
I mean `solveOpts` is pushing back the solver instance on the MStack after it has solved the problem so this could lead to this problem if the output function of type `Problem domain res -> res` were called several times, but I don't see why this would happen.

Reply:
But if that can be a problem I guess trying to free the solver instance is out of the question so I'll just give up on it for now.